File: sample-domain/npm/src/abstractions/repository.ts
import { IAggregate } from './aggregate';

export interface IRepository<TAggregate extends IAggregate<TKey>, TKey=number> {
    getAsync(): Promise<TAggregate[]>;
    getAsync(skip: number, take: number): Promise<TAggregate[]>;
    getAsync(id: TKey): Promise<TAggregate | null>;
    insertAsync(record: TAggregate): Promise<TAggregate>;
    updateAsync(id: TKey, record: TAggregate): Promise<TAggregate>;
    deleteAsync(id: TKey): Promise<void>;
}

File: sample-domain/dotnet/src/abstractions/repository.cs
using System.Collections.Generic;
using System.Threading.Tasks;

namespace Sample.Domain.Abstractions 
{
    public interface IRepository<TAggregate> : IRepository<TAggregate, int> where TAggregate : IAggregate { }
    public interface IRepository<TAggregate, TKey> where TAggregate : IAggregate<TKey> 
    {
        Task<IEnumerable<TAggregate>> GetAsync();
        Task<IEnumerable<TAggregate>> GetAsync(int skip, int take);
        Task<TAggregate> GetAsync(TKey id);
        Task<TAggregate> InsertAsync(TAggregate record);
        Task<TAggregate> UpdateAsync(TKey id, TAggregate record);
        Task DeleteAsync(TKey id);
    }
}